src/crypto/equihash.cpp

```diff
@@ -244,43 +274,271 @@ std::set<std::vector<eh_index>> Equihash::BasicSolve(const eh_HashState& base_st
+            LogPrint("pow", "- Finding collisions\n");
+            for (int i = 0; i < Xt.size() - 1; i++) {
+                TruncatedStepRow<FinalTruncatedWidth> res(Xt[i], Xt[i+1], hashLen, lenIndices, 0);
+                if (res.IsZero(hashLen)) {
+                    partialSolns.push_back(res.GetTruncatedIndices(hashLen, 2*lenIndices));
+                }
+            }
+        } else
+            LogPrint("pow", "- List is empty\n");
+
+    } // Ensure Xt goes out of scope and is destroyed
+
+    LogPrint("pow", "Found %d partial solutions\n", partialSolns.size());
+
+    // Now for each solution run the algorithm again to recreate the indices
+    LogPrint("pow", "Culling solutions\n");
+    std::set<std::vector<eh_index>> solns;
+    eh_index recreate_size { UntruncateIndex(1, 0, CollisionBitLength + 1) };
+    int invalidCount = 0;
+    for (eh_trunc* partialSoln : partialSolns) {
+        // 1) Generate first list of possibilities
+        size_t hashLen = N/8;
+        size_t lenIndices = sizeof(eh_index);
+        std::vector<std::vector<FullStepRow<FinalFullWidth>>> X;
+        X.reserve(soln_size);
+        for (eh_index i = 0; i < soln_size; i++) {
+            std::vector<FullStepRow<FinalFullWidth>> ic;
+            ic.reserve(recreate_size);
+            for (eh_index j = 0; j < recreate_size; j++) {
+                eh_index newIndex { UntruncateIndex(partialSoln[i], j, CollisionBitLength + 1) };
+                ic.emplace_back(N, base_state, newIndex);
+            }
+            X.push_back(ic);
+        }
+
+        // 3) Repeat step 2 for each level of the tree
+        for (int r = 0; X.size() > 1; r++) {
+            std::vector<std::vector<FullStepRow<FinalFullWidth>>> Xc;
+            Xc.reserve(X.size()/2);
+
+            // 2a) For each pair of lists:
+            for (int v = 0; v < X.size(); v += 2) {
+                // 2b) Merge the lists
+                std::vector<FullStepRow<FinalFullWidth>> ic(X[v]);
+                ic.reserve(X[v].size() + X[v+1].size());
+                ic.insert(ic.end(), X[v+1].begin(), X[v+1].end());
+                std::sort(ic.begin(), ic.end(), CompareSR(hashLen));
+                CollideBranches(ic, hashLen, lenIndices, CollisionByteLength, CollisionBitLength + 1, partialSoln[(1<<r)*v], partialSoln[(1<<r)*(v+1)]);
+
+                // 2v) Check if this has become an invalid solution
+                if (ic.size() == 0)
+                    goto invalidsolution;
+
+                Xc.push_back(ic);
+            }
+
+            X = Xc;
+            hashLen -= CollisionByteLength;
+            lenIndices *= 2;
+        }
+
+        // We are at the top of the tree
+        assert(X.size() == 1);
+        for (FullStepRow<FinalFullWidth> row : X[0]) {
+            solns.insert(row.GetIndices(hashLen, lenIndices));
+        }
+        goto deletesolution;
+
+invalidsolution:
+        invalidCount++;
+
+deletesolution:
+        delete[] partialSoln;
```

Comment:
Isn't this automatically deleted when it goes out of scope or is assigned the next element of ``partialSolns``?